solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -1802,6 +1800,18 @@ public void waitForState(
       throw new AlreadyClosedException();
     }
 
+    // Check predicate against known clusterState before trying to add watchers
+    if (clusterState != null) {
+      Set<String> liveNodes = clusterState.getLiveNodes();
+      DocCollection docCollection = clusterState.getCollectionOrNull(collection);
+      if (liveNodes != null && docCollection != null) {
+        if (predicate.matches(liveNodes, docCollection)) {
+          log.debug("Found {} directly in clusterState", predicate);
```

Comment:
Probably the `toString()` output of the predicate won't be relevant most of the time. I'm not sure what is the usual practice in Solr for this.

Reply:
Hmmm so I see what you are saying @psalagnac. I took inspiration from line 1862 where log.debug is doing basically the same logic in the other wait for state.

Reply:
I'm going to leave this as is for now.